index.d.ts

```diff
@@ -1,288 +1,591 @@
-    manualSorting?: boolean
-    disableSorting?: boolean
-    defaultSortDesc?: boolean
-    disableMultiSort?: boolean
-  }
+export type UseGroupByOptions<D extends object> = Partial<{
+  manualGroupBy: boolean
+  disableGrouping: boolean
+  aggregations: Record<string, AggregatorFn<D>>
+  groupByFn: (rows: Row<D>[], columnId: IdType<D>) => Record<string, Row<D>>
+}>
+
+export interface UseGroupByHooks<D extends object> {
+  getGroupByToggleProps: ((
+    header: HeaderGroup<D>,
+    instance: TableInstance<D>
+  ) => object)[]
+}
 
-  export interface UseSortbyState<D> {
-    sortBy?: SortingRules<D>
-  }
+export interface UseGroupByState<D extends object> {
+  groupBy: IdType<D>[]
+}
 
-  export interface TableInstance<D = {}> extends TableOptions<D> {
-    hooks: Hooks
-    rows: Row<D>[]
-    columns: EnhancedColumn<D>[]
-    headerGroups: HeaderGroup<D>[]
-    headers: HeaderGroup<D>[]
-    getTableProps: (userProps?: any) => any
-    getRowProps: (userProps?: any) => any
-    prepareRow: (row: Row<D>) => any
-  }
+export type UseGroupByColumnOptions<D extends object> = Partial<{
+  aggregate: Aggregator<D> | Aggregator<D>[]
+  Aggregated: Renderer<CellProps<D>>
+  disableGrouping: boolean
+  groupByBoundary: boolean
+}>
+
+export interface UseGroupByInstanceProps<D extends object> {
+  rows: Row<D>[]
+  preGroupedRows: Row<D>[]
+  toggleGroupBy: (columnId: IdType<D>, toggle: boolean) => void
+}
 
-  export interface TableOptions<D = {}> {
-    data: D[]
-    columns: HeaderColumn<D>[]
-    state: State<D>
-    debug?: boolean
-    loading: boolean
-    defaultColumn?: Partial<Column<D>>
-  }
+export interface UseGroupByColumnProps<D extends object> {
+  canGroupBy: boolean
+  isGrouped: boolean
+  groupedIndex: number
+  toggleGroupBy: () => void
+}
+
+export interface UseGroupByHeaderProps<D extends object> {
+  getGroupByToggleProps: (props?: object) => object
+}
+
+export interface UseGroupByRowProps<D extends object> {
+  isAggregated: boolean
+  groupByID: IdType<D>
+  groupByVal: string
+  values: Record<IdType<D>, AggregatedValue>
+  subRows: Row<D>[]
+  depth: number
+  path: IdType<D>[]
+  index: number
+}
+
+export interface UseGroupByCellProps<D extends object> {
+  isGrouped: boolean
+  isRepeatedValued: boolean
+  isAggregated: boolean
+}
+
+export type DefaultAggregators =
+  | 'sum'
+  | 'average'
+  | 'median'
+  | 'uniqueCount'
+  | 'count'
+
+export type AggregatorFn<D extends object> = (
+  columnValues: CellValue[],
+  rows: Row<D>[]
+) => AggregatedValue
+export type Aggregator<D extends object> =
+  | AggregatorFn<D>
+  | DefaultAggregators
+  | string // TODO: Check that DefaultAggregators works
+export type AggregatedValue = unknown
+/* #endregion */
+
+/* #region usePagination */
+export function usePagination<D extends object = {}>(hooks: Hooks<D>): void
+export namespace usePagination {
+  const pluginName = 'usePagination'
+}
+
+export type UsePaginationOptions<D extends object> = Partial<{
```

Comment:
based on https://github.com/Microsoft/TypeScript/wiki/FAQ#why-is-astring-assignable-to-anumber-for-interface-at-- I wonder if we shouldn't be more careful about not using generic signatures on interfaces that aren't generic.  I know that it makes using the interface harder since you have to be careful about when you do and don't pass in generic options, but when I've got this wrong in the past I've ended out with complex interfaces that ended out allowing pretty much any old garbage through, i.e. they resolved to `any`.

Reply:
Other than some slight confusion of if they pass `UsePaginationOptions<Garbage>` to something they previously defined as `UsePaginationOptions<Person>`, I only see a downside in confusion as to when to use and not use the generics. It shouldn't ever resolve to `any` unless you are unioning it with `any`. 

In general, I wouldn't add the generic for cases where we don't need it. However, in our case, where the developers are having to do a bunch of interface merging, I think it makes sense to leave it until we can clean up the usage of these types.

Reply:
My concern is that I've seen cases where doing this just breaks the interface, and does so in rather hard to find ways.  When your type suddenly becomes an effective alias for any, your code still compiles, you've just lost the compilation changes that you thought you had.

I'll have to dig more and see if I can come up with a better example.  I'm all for simplicity, but it has to actually work, and when the docs say "don't do the thing", it seemed wise to consider what they say.

Reply:
I'd say that this is good as is.  We can change it later if there's a discovered problem.